### PR TITLE
truncate strings that are stored as keywords in ES (#159)

### LINKED ISSUE #156: Problem validating JSON document against schema - length must be <= 1024, but got 1523
Continuously seeing these errors in the log -

2018-02-13T05:06:49.104Z	ERROR	[request]	beater/handlers.go:384	error handling request	{"request_id": "79f54c15-f99e-4d1d-ab14-07bf622983c8", "error": "Problem validating JSON document against schema: I[#] S[#] doesn't validate with \"transaction#\"\n  I[#/transactions/29/context/request/url/full] S[#/properties/transactions/items/properties/context/properties/request/properties/url/properties/full/maxLength] length must be <= 1024, but got 1523"}

apm-server version is 6.2.1
elastic-apm agent version 2.0.0 with Flask.

Is there any config property that I need to change to let the apm server accept urls of a longer length?

### PR DESCRIPTION
**Pull request by beniwohli**
this should avoid schema errors for values that are too long

closes #156

closes #159

### CODE REVIEW

**Comment by beniwohli** [comment]
jenkins test this

File: elasticapm/utils/encoding.py [hunk 1]
```diff
@@ -165,3 +168,9 @@ def shorten(var, list_length=50, string_length=200):
         # TODO: when we finish the above, we should also implement this for dicts
         var = list(var)[:list_length] + ['...', '(%d more elements)' % (len(var) - list_length,)]
     return var
+
+
+def keyword_field(string):
+    if not isinstance(string, compat.string_types) or len(string) <= KEYWORD_MAX_LENGTH:
+        return string
+    return string[:KEYWORD_MAX_LENGTH - 1] + u'…'
```
Comment (jahtalab) [inline elasticapm/utils/encoding.py]: @beniwohli , Adding '…' would add 3 characters to the length which might exceed 1024 limit.

A separate issue: In my opinion '...' is not necessary, I would prefer adding a property to say the name was truncated and let the ui decide about the "ellipsis", but even that I think at this stage is not necessary.

File: elasticapm/base.py [hunk 2]
```diff
@@ -327,45 +328,48 @@ def _send_remote(self, url, data, headers=None):
...
             }
+        self._service_info = result
         return result
 
     def get_process_info(self):
         return {
             'pid': os.getpid(),
             'argv': sys.argv,
-            'title': None,
+            'title': None,  # Note: if we implement this, the value needs to be wrapped with keyword_field
```
Comment (jahtalab) [inline elasticapm/base.py]: I like your note to your future self 😄 , I do that as well!

File: elasticapm/utils/encoding.py (hunk 1 above)
Comment (beniwohli) [reply elasticapm/utils/encoding.py]: The character I use there is the unicode ellipsis, so it's only one character :)

I'm not sure adding an additional field for every keyword field to indicate if it was truncated or not is feasible. Having to truncate a field should be an absolute edge case.

Comment (jahtalab) [reply elasticapm/utils/encoding.py]: I see, sorry my python is a bit rusty 😄 !


It doesn't have to be a field for every keyword field. But my point was that I would prefer not adding anything to the fields that might have a meaning in different contexts and keeping what the data was originally

Comment (jahtalab) [reply elasticapm/utils/encoding.py]: I'm going to approve the changes, then you can decide about the 'ellipsis'!

**Pull request closed without merging**